subsys/net/lib/tftp/tftp_client.c

```diff
@@ -0,0 +1,306 @@
+/*
+ * Copyright (c) 2020 InnBlue
+ *
+ * SPDX-License-Identifier: Apache-2.0
+ */
+
+#include <logging/log.h>
+LOG_MODULE_REGISTER(tftp_client, LOG_LEVEL_DBG);
+
+/* Public / Internal TFTP header file. */
+#include <net/tftp.h>
+#include "tftp_client.h"
+
+/* TFTP Global Buffer. */
+static u8_t   tftpc_buffer[TFTPC_MAX_BUF_SIZE];
```

Comment:
I'd consider this bad design, this locks down non-trivial amount of memory (512+ bytes) for TFTP usage, which may never even happen. The way to solve this issue is to follow "dependency injection" pattern, where buffers are allocated and managed by the application, and `struct tftpc` contains a pointer to this buffer. This way, the same buffer can be used, in time-shared manner, for TFTP, HTTP, app's own processing, etc.

For reference, we have an RFC for guidelines on how to implement higher-level networking protocols: https://github.com/zephyrproject-rtos/zephyr/issues/13212 . This particular matter might not be there, but the point is that we do have some (proposed) guidelines.

Bottom line: I don't consider this to be worth -1, as TFTP is a special protocol, which would be used only for special needs (e.g. OTA-like). But I suggest to queue this up for a future refactor.

Reply:
I had a couple of options when I started thinking about this problem, 

- **malloc buffer on user request** - When the user calls TFTP to get or put files from the server, we can dynamically allocate / deallocate buffers on every request. As this can cause memory fragmentation very quickly (if we have large number of tftp requests), this was not a very good solution. 
- **allocate buffers on the stack** - Instead of having a "global" internal buffer, we create a "512+4" bytes sized buffers on the stack but this imposes unnecessary burden on each calling task stack, I went against it as well.. 

Given all this, I decided to keep a global buffer which would get filled and eventually copied into the user buffer. I understand the problems that this approach has but it was simpler to implement and so I opted for this over the approach you've recommended here.. 

I had thought about implementing the approach mentioned by you a couple of weeks ago, but it requires "some" extra work and that's why I considered the current implementation to be okay for this version. Eventually we want to get rid of this "global" buffer as it is consuming large amount of memory and implement the mentioned approach..

Reply:
> I had thought about implementing the approach mentioned by you a couple of weeks ago, but it requires "some" extra work

Yes, and the proposed solution to this is to have "baseline" API which adheres to the best practices of e.g. resource usage, widest feature coverage, and composability, and on top of that, have an optional "convenience API", which has less knobs to control, and thus easier to use, at the expense of efficiency.

Example of that design is e.g. SNTP API: you can either call individual lower-level functions which allow you to have full control of what happens, or use a higher-level wrapper `sntp_simple()` https://github.com/zephyrproject-rtos/zephyr/blob/master/include/net/sntp.h#L98 which calls a number of functions for you. This way, we allow to develop with Zephyr both highly optimized and customized solutions, and at the same, code up typical usage scenarios quickly.

As I mentioned, I don't insist on this refactor for this PR (though it's pretty simple, and the drawback of not doing it now is breaking API changes later).

Reply:
your feedback is very valuable, in my opinion -- it is better to invest some time now, instead of taking that backpack with us and have a mountain to pass later. 
let us mitigate that part.

thanks!

Reply:
> I had a couple of options when I started thinking about this problem,
> 
>     * **malloc buffer on user request** - When the user calls TFTP to get or put files from the server, we can dynamically allocate / deallocate buffers on every request. As this can cause memory fragmentation very quickly (if we have large number of tftp requests), this was not a very good solution.
> 
>     * **allocate buffers on the stack** - Instead of having a "global" internal buffer, we create a "512+4" bytes sized buffers on the stack but this imposes unnecessary burden on each calling task stack, I went against it as well..

There is a third option, just leave the allocations headache to the caller of this library. So the API would require that the caller supplies any extra buffers and lengths. This way we avoid extra management of buffers and stack usage, and application developer can optimize things in the application side. I think we discussed this issue earlier but you wanted to postpone this kind of change to later. Best of course would be to do that now as we have not yet released the API. The API is marked experimental so we can change it more easily later too.

Paul had a good point that we would need a sample app that just shows how the API can be used, so +1 to that.